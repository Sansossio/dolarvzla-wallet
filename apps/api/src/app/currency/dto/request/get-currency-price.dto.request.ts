import { ApiProperty } from '@nestjs/swagger'
import { IsEnum, IsNotEmpty } from 'class-validator'
import { Currency } from '@dolarvzla-wallet/models'

export class GetCurrencyPriceRequestDto {
  @ApiProperty({
    enum: Currency,
    example: Currency.BITCOIN
  })
  @IsEnum(Currency)
  @IsNotEmpty()
  currency: Currency
}
