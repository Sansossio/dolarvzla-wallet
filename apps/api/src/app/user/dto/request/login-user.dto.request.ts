import { ApiProperty } from '@nestjs/swagger'
import { IsEmail, IsNotEmpty, IsString } from 'class-validator'

export class LoginUserDtoRequest {
  @ApiProperty({
    example: 'email@example.com'
  })
  @IsEmail()
  @IsNotEmpty()
  email: string

  @ApiProperty()
  @IsString()
  @IsNotEmpty()
  password: string
}
