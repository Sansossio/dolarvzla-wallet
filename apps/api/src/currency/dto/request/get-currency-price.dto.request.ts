import { ApiProperty } from '@nestjs/swagger'
import { IsEnum, IsNotEmpty } from 'class-validator'
import { Currency } from '../../enum/currency.enum'

export class GetCurrencyPriceRequestDto {
  @ApiProperty({
    enum: Currency,
    example: Currency.BITCOIN
  })
  @IsEnum(Currency)
  @IsNotEmpty()
  currency: Currency
}
