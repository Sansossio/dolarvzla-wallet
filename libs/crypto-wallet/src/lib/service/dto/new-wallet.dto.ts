export class NewWalletDto {
  name: string
  address: string
}
