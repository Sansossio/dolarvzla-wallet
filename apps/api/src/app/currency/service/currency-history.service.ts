import { Injectable, NotFoundException } from '@nestjs/common'
import { InjectRepository } from '@nestjs/typeorm'
import { Repository } from 'typeorm'
import { CurrencyHistoryEntity } from '@dolarvzla-wallet/models'
import { GetCurrencyPriceRequestDto } from '../dto/request/get-currency-price.dto.request'
import { GetCurrencyPriceDtoResponse } from '../dto/response/get-currency-price.dto.response'

@Injectable()
export class CurrencyHistoryService {
  constructor (
    @InjectRepository(CurrencyHistoryEntity)
    private readonly repository: Repository<CurrencyHistoryEntity>
  ) {}

  async getCurrencyPrice (query: GetCurrencyPriceRequestDto): Promise<GetCurrencyPriceDtoResponse> {
    const data = await this.repository.findOne({
      where: {
        currency: query.currency
      },
      order: {
        createAt: 'DESC'
      },
      select: [
        'price',
        'createAt'
      ]
    })

    if (!data) {
      throw new NotFoundException()
    }

    return {
      currency: query.currency,
      price: data.price,
      lastUpdate: data.createAt
    }
  }
}
