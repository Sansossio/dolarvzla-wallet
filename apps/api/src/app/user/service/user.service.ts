import { ConflictException, Injectable } from '@nestjs/common'
import { InjectRepository } from '@nestjs/typeorm'
import { Repository } from 'typeorm'
import { UserEntity } from '@dolarvzla-wallet/models'
import { RegisterUserDtoRequest } from '../dto/request/register-user.dto.request'

@Injectable()
export class UserService {
  constructor (
    @InjectRepository(UserEntity)
    private readonly repository: Repository<UserEntity>
  ) {}

  async register (body: RegisterUserDtoRequest) {
    const exists = !!await this.repository.findOne({
      where: {
        email: body.email
      },
      select: [
        'id'
      ]
    })
    if (exists) {
      throw new ConflictException('User already exists')
    }
    const newUser = this.repository.create({
      email: body.email,
      password: body.password
    })

    await this.repository.save(newUser)
  }
}
