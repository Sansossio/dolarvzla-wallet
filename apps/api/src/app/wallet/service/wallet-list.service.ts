import { Injectable } from '@nestjs/common'
import { Repository } from 'typeorm'
import { UserWalletEntity } from '@dolarvzla-wallet/models'
import { InjectRepository } from '@nestjs/typeorm'
import { WalletListResponseDto } from '../dto/response/wallet-list.response.dto'
import { WalletOrchestrator } from '@dolarvzla-wallet/crypto-wallet'

@Injectable()
export class WalletListService {
  constructor (
    @InjectRepository(UserWalletEntity)
    private readonly repository: Repository<UserWalletEntity>,

    private readonly walletOrquestador: WalletOrchestrator
  ) {}

  async list (userId: number): Promise<WalletListResponseDto[]> {
    const listOfWallets = await this.repository.find({
      where: {
        user: { id: userId }
      },
      select: [
        'type',
        'walletId',
        'address'
      ]
    })

    return Promise.all(
      listOfWallets.map(async (wallet): Promise<WalletListResponseDto> => {
        const { balance } = await this.walletOrquestador
          .getHandler(wallet.type)
          .getBalance(wallet.walletId)

        return WalletListResponseDto.fromEntity(wallet, balance)
      })
    )
  }
}
