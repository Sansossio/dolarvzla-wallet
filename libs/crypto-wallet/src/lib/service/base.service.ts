import { HttpService, Injectable } from '@nestjs/common'
import { ConfigService } from '@nestjs/config'
import { NewWalletDto } from './dto/new-wallet.dto'
import { WalletBalanceDto } from './dto/wallet-balance.dto'

@Injectable()
export abstract class BaseService {
  constructor (
    protected readonly httpService: HttpService,
    protected readonly config: ConfigService
  ) {}

  abstract createRandomWallet (): Promise<NewWalletDto>

  abstract listWallets (): Promise<string[]>

  abstract generateNewAddress (walletName: string): Promise<string>

  abstract getBalance (walletname: string): Promise<WalletBalanceDto>
}
