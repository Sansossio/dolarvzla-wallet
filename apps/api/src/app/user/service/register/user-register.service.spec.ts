import { ConflictException } from '@nestjs/common'
import { RegisterUserDtoRequest } from '../../dto/request/register-user.dto.request'
import { UserRegisterService } from './user-register.service'

describe('User Register service', () => {
  const mockedBody: RegisterUserDtoRequest = {
    email: 'myemail',
    password: 'mypassword'
  }

  const repository = {
    findOne: jest.fn(),
    create: jest.fn(),
    save: jest.fn()
  }

  const btcService = {
    createRandomWallet: jest.fn()
  }

  const service = new UserRegisterService(
    repository as any,
    btcService as any
  )

  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should be defined', () => {
    expect(service).toBeDefined()
  })

  it('should throw conflict error when the user already exists', async () => {
    repository.findOne.mockImplementation(async () => ({ id: 1 }))

    await expect(service.register(mockedBody)).rejects.toBeInstanceOf(ConflictException)
  })

  it('shouldnt create user when create wallet fails', async () => {
    btcService.createRandomWallet.mockImplementation(async () => Promise.reject(new Error()))

    await expect(service.register(mockedBody)).rejects.toBeInstanceOf(Error)

    expect(btcService.createRandomWallet).toBeCalledTimes(1)
    expect(repository.save).toBeCalledTimes(0)
  })

  it('should create user and wallet', async () => {
    btcService.createRandomWallet.mockImplementation(async () => ({ name: 'test', address: 'address' }))
    await expect(service.register(mockedBody)).resolves.toMatchSnapshot()

    expect(btcService.createRandomWallet).toBeCalledTimes(1)
    expect(repository.save).toBeCalledTimes(1)
  })
})
