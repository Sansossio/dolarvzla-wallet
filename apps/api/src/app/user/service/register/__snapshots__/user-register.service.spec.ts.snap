// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Register service should create user and wallet 1`] = `undefined`;
